adapter/stackdriver/stackdriver.go

```diff
@@ -0,0 +1,312 @@
+// Copyright 2017 the Istio Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package stackdriver
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	monitoring "cloud.google.com/go/monitoring/apiv3"
+	"github.com/golang/protobuf/ptypes"
+	"github.com/googleapis/gax-go"
+	xcontext "golang.org/x/net/context"
+	gapiopts "google.golang.org/api/option"
+	labelpb "google.golang.org/genproto/googleapis/api/label"
+	metricpb "google.golang.org/genproto/googleapis/api/metric"
+	"google.golang.org/genproto/googleapis/api/monitoredres"
+	monitoringpb "google.golang.org/genproto/googleapis/monitoring/v3"
+
+	"istio.io/mixer/adapter/stackdriver/config"
+	"istio.io/mixer/pkg/adapter"
+)
+
+// TODO: implement adapter validation
+// TODO: change batching to be size aware: right now we batch and send data to stackdriver based on only a ticker.
+// Ideally we'd also size our buffer and send data whenever we hit the size limit or config.push_interval time has passed
+// since the last push.
+// TODO: today we start a ticker per aspect instance, each keeping an independent data set it pushes to SD. This needs to
+// be promoted up to the factory, which will hold a buffer that all aspects write in to, with a single ticker/loop responsible
+// for pushing the data from all aspect instances.
+
+type (
+
+	// Abstracts the creation of the stackdriver client to enable network-less testing.
+	createClientFunc func(*config.Params) (*monitoring.MetricClient, error)
+
+	factory struct {
+		adapter.DefaultBuilder
+		createClient createClientFunc
+
+		once sync.Once
+		t    *time.Ticker // we hold on to a ref so we can stop it during Close()
```

Comment:
super-nit: i might be tempted to call this `tick` (to match `once`), but up to you.